### ISSUE TITLE
Wish: zmq_getsockopt() return correct property size when insufficient size given

### ISSUE THREAD

**Issue by hostilefork** [zeromq/libzmq#3160]
# Issue description

Typically, APIs which take an address to give back a "real size" for what a property is can be called in two passes.  The first pass gives a small size (possibly even 0 with a NULL buffer pointer) with the intent of receiving the actual size.  This allows the caller to then allocate an exactly sized buffer, which in a second pass receives the data.
 
`zmq_getsockopt()`, however, fails without informing the caller of the size.  It knows what it is, because it just checked against it:

https://github.com/zeromq/libzmq/blob/6bfa91f13f73b67be47ded18986b76d154d5923b/src/options.cpp#L76

From the remarks in the code (which may or may not apply here), it could even be impossible to "guess" at the size.  Exact buffer sizes might be required, and gradually increasing the buffer size may pass over that size.

It would seem more informative if:

    if (*optvallen_ < value_len_) {
        return sockopt_invalid ();
    }

Were changed to something like:

    if (*optvallen_ < value_len_) {
        *optvallen_ = value_len_;
        return sockopt_invalid ();
    }

...although the fact that there's a ZMQ_ACT_MILITANT mode triggered by sockopt_invalid() suggests that a distinct return result might be required.

# Environment

* libzmq version: 4.1.7 (and presumably all prior)
* OS: all

**Comment by sigiesec** [comment]
This could probably be done. However, I am not sure why you would need this. The required resp. maximum size of all socket options is documented. For what purpose would you use the result?

**Comment by hostilefork** [comment]
> The required resp. maximum size of all socket options is documented.

Where is this documented?  I don't see anything on the [zmg_getsockopt()](http://api.zeromq.org/4-1:zmq-getsockopt) page which defines the maximum size for, say, ZMQ_PLAIN_USERNAME.

> For what purpose would you use the result?

In a wrapper, which internally has a table mapping the options to their types, and generates dynamic values without the user needing to pass in a buffer.  Even with a maximum size known, it would be wasteful if e.g. a string option were a smaller size to allocate such a large buffer.

> It's also a change in the existing API, so we can't really do it

Regardless, it's useful to [have a place to link to](http://blog.hostilefork.com/comments-vs-web-links-2014/) about it (for me or anyone else facing the problem).

**Comment by sigiesec** [comment]
Let me rephrase this:
> The required resp. maximum size of all socket options _should be_ documented.

In the case of ZMQ_PLAIN_USERNAME, the maximum size is 256 bytes. This is indeeed missing from the documentation, and should be added.

> Even with a maximum size known, it would be wasteful if e.g. a string option were a smaller size to allocate such a large buffer if the value were much smaller.

In general this might be required, but I think there are no options in libzmq that could become large (larger than 256 bytes).

**Comment by bluca** [comment]
@hostilefork could you please send a PR to update that option's doc?

**Comment by hostilefork** [comment]
@bluca It's not within my scope of knowledge/authority to say that's a fact...just hearsay as far as I know.  :-)  And there's nothing in particular about that one particular option that interests me (it was just an example)...so really it would take more research to go through and validate the claims/etc.

I'll leave it up to the powers that be to explore it further, or close the issue...

**Comment by ADDubovik** [comment]
The behaviour of zmq_getsockopt is really need to be corrected as topicstarter says.
